### PR TITLE
Auth/msarps

### PR DESCRIPTION
**Pull request by satyarapelly**
#### What does this PR do? Please provide some context
This PR, invoke or direct the MSA-RPS API once user is registered using Live authentication.
MSA-API redirection once new user is registered to get the PUID for logged-in user browser context (single sign-in), invoke the openedx rest api using LTI consumer approach., request will be validated using LTI signature that is generated using secret key at the consumer side. Once request been validated, API will create the social auth mapping table using uid, puid.

#### Where should the reviewer start?
common/djangoapps/third_party_auth/models.py
common/djangoapps/student/views.py

#### How can this be manually tested? (brief repro steps)
1. This required C# WebAPI hosted on the local machine is part of LEX TFS location. $/MLX/Prototypes/LR.ExternalLogin.Web
2.  Admin settings as follows:
     i. Create the LTI Consumer entry 
     ii. Configure the same secret key in C# web api's web.config
     iii.  Add extra Site_configurations entries
{
  
  "external_login_api":"https://msa.rps.signing.microsoft.com/api/v1.0/ExternalLogin/Authenticate?Provider=RPS&RedirectUri=",
   "SOCIAL_OAUTH_MSA_PROVIDER":"oa2-live"
}
#### What are the relevant TFS items? (list id numbers)
All the tasks under the story: 102365
#### Definition of done:
- [x] Title of the pull request is clear and informative
- [x] Add pull request hyperlink to relevant TFS items
- [x] For large or complex change: schedule an in-person review session
- [x] This change has appropriate test coverage

#### Reminders BEFORE merging
1. Get at least two approvals
1. If you're merging from a feature branch into the development branch then "flatten" or "squash" commits
1. If merging from the development branch into master (or porting changes from upstream) then use github's UI to get review feedback, but use the git command line interface to complete the actual merge.

#### Reminders AFTER merging
1. Delete the remote feature branch
1. Resolve relevant TFS items
1. (reverse merge) If you merged from the development branch into master then check to see if there are any changes in master that can be merged down to the development branch (like hotfixes, etc). In this case, use github's UI for feedback and the git command line interface for the actual merge.

[//]: # ( todo: If you merged into development branch then verify change in our "rolling deployment" environment. Then notify stakeholders interested in or involved with the change )

[//]: # ( fyi: This content was heavily inspired by )
[//]: # ( 1 Our team's policies and processes )
[//]: # ( 2 https://github.com/sprintly/sprint.ly-culture/blob/master/pr-template.md )
[//]: # ( 3 The book "The Checklist Manifesto: How to Get Things Right" by Atul Gawande )
[//]: # ( 4 https://github.com/Azure/azure-event-hubs/blob/master/.github/PULL_REQUEST_TEMPLATE.md )

### CODE REVIEW

File: common/djangoapps/student/helpers.py [hunk 1]
```diff
@@ -221,7 +221,7 @@ def auth_pipeline_urls(auth_entry, redirect_url=None):
 
     return {
         provider.provider_id: third_party_auth.pipeline.get_login_url(
-            provider.provider_id, auth_entry, redirect_url=redirect_url
+            provider.provider_id, auth_entry, redirect_url=redirect_url + "?provider={}".format(provider.provider_id)
```
Comment (sdolenc) [inline common/djangoapps/student/helpers.py]: please us a url builder instead of concatenating strings. advantage over the current approach:
- what if the url already contains a query paramater? the result would be two question marks in the url (whereas the second one should be an ampersand)
- url builders generally help with "url" encoding too

File: lms/envs/aws.py [hunk 2]
```diff
@@ -668,6 +668,7 @@
             'social.backends.linkedin.LinkedinOAuth2',
             'social.backends.facebook.FacebookOAuth2',
             'social.backends.azuread.AzureADOAuth2',
+            'social.backends.live.LiveOAuth2',
```
Comment (sdolenc) [inline lms/envs/aws.py]: I strongly recommend we take #115 for this change instead. it's the same change to aws.py, but isolated. That way it can be cherry-picked when we're deploy standard upstream edx (without our customization). For more information: see the todo in enable_msft_oauth.sh within https://github.com/Microsoft/oxa-tools/pull/219/files

Please approve #115 when you get the chance so we can merge my fix before this larger change. #120

File: common/djangoapps/third_party_auth/views.py [hunk 3]
```diff
@@ -10,7 +10,8 @@
 from social.apps.django_app.views import complete
 from social.apps.django_app.utils import load_strategy, load_backend
 from social.utils import setting_name
-from .models import SAMLConfiguration
+from .models import SAMLConfiguration, UserSocialAuthMapping
+from social.apps.django_app.default.models import UserSocialAuth
```
Comment (sdolenc) [inline common/djangoapps/third_party_auth/views.py]: i noticed you updated the imports, but don't use it within the file. these might be dumb questions, but i figured i'd ask:
1. are these changes required? in other words, will everything keep working if reverted?
2. if not, can these imports be moved to the file where the libraries are used? (putting the imports closer to their usage?)

File: common/djangoapps/student/helpers.py (hunk 1 above)
Comment (satyarapelly) [reply common/djangoapps/student/helpers.py]: Yes, but we are not directly using to redirect it, if you see, this does not have whole url , instead of it just holds the value -  /dashboard?provider=live. And consumer will actually do the url encode.

File: common/djangoapps/third_party_auth/views.py (hunk 3 above)
Comment (satyarapelly) [reply common/djangoapps/third_party_auth/views.py]: Perfect, i have changed my mind, moved it to different place.

File: common/djangoapps/student/helpers.py (hunk 1 above)
Comment (sdolenc) [reply common/djangoapps/student/helpers.py]: url builders can handle relative paths without domains. we should absolutely use one

my broader point: we should never manipulate urls like they're simple strings (parsing, concatenating, etc may work in some cases, but it's brittle and insecure in the long run) . we should always use url builders or wrapper functions that take into account a ton of edge cases for us. python is a super rich language with lots of of options to get this done easily. i'm sure there are cases of where edx is doing this properly we can borrow form

[the remark this answers is not in the document]
Comment (satyarapelly) [reply common/djangoapps/student/helpers.py]: No, we have condition to handle the provider is oa2-live, rest of the cases we ignore.

File: common/djangoapps/student/helpers.py [hunk 4]
```diff
@@ -221,7 +221,7 @@ def auth_pipeline_urls(auth_entry, redirect_url=None):
 
     return {
         provider.provider_id: third_party_auth.pipeline.get_login_url(
-            provider.provider_id, auth_entry, redirect_url=redirect_url
+            provider.provider_id, auth_entry, redirect_url="{}?{}".format(redirect_url, urllib.urlencode({"provider": provider.provider_id}))
```
Comment (sdolenc) [inline common/djangoapps/student/helpers.py]: this doesn't solve the problem i have, but i'll unblock
